File: CamStudy/CamStudy/ViewModels/StudyRoomListViewModel.swift
import FirebaseFirestore
import Foundation

@MainActor
final class StudyRoomListViewModel: ObservableObject {
    @Published var rooms: [StudyRoom] = []
    @Published var errorMessage: String?

    private var listener: ListenerRegistration?
    private let firestoreService = FirestoreService.shared

    func startObserving() {
        guard listener == nil else { return }
        listener = firestoreService.observeStudyRooms { [weak self] result in
            Task { @MainActor in
                guard let self else { return }
                switch result {
                case .success(let rooms):
                    self.rooms = rooms
                case .failure(let error):
                    self.errorMessage = error.localizedDescription
                }
            }
        }
    }

    func stopObserving() {
        listener?.remove()
        listener = nil
    }
}

File: camstudy_app/lib/viewmodels/study_room_list_view_model.dart
import 'dart:async';

import 'package:flutter/foundation.dart';

import '../models/study_room.dart';
import '../services/firestore_service.dart';

class StudyRoomListViewModel extends ChangeNotifier {
  List<StudyRoom> rooms = [];
  String? errorMessage;

  StreamSubscription<List<StudyRoom>>? _subscription;
  final _firestoreService = FirestoreService.shared;

  void startObserving() {
    if (_subscription != null) return;
    _subscription = _firestoreService.observeStudyRooms().listen(
      (rooms) {
        this.rooms = rooms;
        notifyListeners();
      },
      onError: (Object error) {
        errorMessage = error.toString();
        notifyListeners();
      },
    );
  }

  void stopObserving() {
    _subscription?.cancel();
    _subscription = null;
  }

  @override
  void dispose() {
    stopObserving();
    super.dispose();
  }
}
